pkg/controller/populators/populator-base.go

```diff
@@ -0,0 +1,335 @@
+		return nil
+	}
+
+	// Rebind PVC to target PVC
+	pv.Annotations = make(map[string]string)
+	pv.Spec.ClaimRef = &corev1.ObjectReference{
+		Namespace:       targetPVC.Namespace,
+		Name:            targetPVC.Name,
+		UID:             targetPVC.UID,
+		ResourceVersion: targetPVC.ResourceVersion,
+	}
+	r.log.V(3).Info("Rebinding PV to target PVC", "PVC", targetPVC.Name)
+	if err := r.client.Update(context.TODO(), pv); err != nil {
+		return err
+	}
+
+	return nil
+}
+
+func (r *ReconcilerBase) reconcile(req reconcile.Request, populator populatorController, log logr.Logger) (reconcile.Result, error) {
+	// Get the target PVC
+	pvc := &corev1.PersistentVolumeClaim{}
+	if err := r.client.Get(context.TODO(), req.NamespacedName, pvc); err != nil {
+		if k8serrors.IsNotFound(err) {
+			return reconcile.Result{}, nil
+		}
+		return reconcile.Result{}, err
+	}
+
+	// We first perform the common reconcile steps.
+	// We should only continue if we get a valid PVC'
+	pvcPrime, err := r.reconcileCommon(pvc, populator, log)
+	if err != nil {
+		return reconcile.Result{}, err
+	}
+	if pvcPrime == nil {
+		return reconcile.Result{}, nil
+	}
+
+	// Each populator reconciles the target PVC in a different way
+	if cc.IsUnbound(pvc) {
+		return populator.reconcileTargetPVC(pvc, pvcPrime)
+	}
+
+	// Making sure to clean PVC'
+	return r.reconcileCleanup(pvcPrime)
+}
+
+func (r *ReconcilerBase) reconcileCommon(pvc *corev1.PersistentVolumeClaim, populator populatorController, log logr.Logger) (*corev1.PersistentVolumeClaim, error) {
+	// We should ignore PVCs that aren't for this populator to handle
+	dataSourceRef := pvc.Spec.DataSourceRef
+	if dataSourceRef == nil || !IsPVCDataSourceRefKind(pvc, r.sourceKind) || dataSourceRef.Name == "" {
+		log.V(1).Info("reconciled unexpected pvc, ignoring")
+		return nil, nil
+	}
+	// Wait until dataSourceRef exists
+	populationSource, err := populator.getPopulationSource(pvc.Namespace, dataSourceRef.Name)
+	if populationSource == nil {
+		return nil, err
+	}
+	ready, waitForFirstConsumer, err := r.handleStorageClass(pvc)
+	if !ready || err != nil {
+		return nil, err
+	}
+
+	// Get the PVC'
+	pvcPrime, err := r.getPVCPrime(pvc)
+	if err != nil {
+		return nil, err
+	}
+
+	// If PVC' doesn't exist and target PVC is not bound, we should create the PVC' to start the population.
+	// We still return the nil PVC' as we'll get called again once PVC' exists.
+	// If target PVC is bound, we don't really need to populate anything.
+	if pvcPrime == nil && cc.IsUnbound(pvc) {
+		_, err := r.createPVCPrime(pvc, populationSource, waitForFirstConsumer, populator.updateAnnotations)
+		if err != nil {
+			return nil, err
+		}
+	}
+
+	return pvcPrime, nil
+}
+
+func (r *ReconcilerBase) reconcileCleanup(pvcPrime *corev1.PersistentVolumeClaim) (reconcile.Result, error) {
+	if pvcPrime != nil {
+		if err := r.client.Delete(context.TODO(), pvcPrime); err != nil {
+			return reconcile.Result{}, err
+		}
+	}
+	return reconcile.Result{}, nil
```

Comment:
why delete the success event I had here?

Reply:
I moved it to `reconcileTargetPVC` so we can use specific events depending on the populator. Once the target PVC is bound I think it's safe to trigger the event.